torch/testing/_internal/common_methods_invocations.py

```diff
@@ -74,30 +75,69 @@ def __init__(self, cls_name=None, test_name=None, *,
                          test_name=test_name, device_type=device_type, dtypes=dtypes,
                          active_if=active_if)
 
+
 class SampleInput(object):
     """Represents sample inputs to a function."""
 
-    # output_process_fn_grad is a function that modifies the output of op compatible with input
-    __slots__ = ['input', 'args', 'kwargs', 'output_process_fn_grad']
+    __slots__ = ['input', 'args', 'kwargs', 'output_process_fn_grad', 'ref', 'extra']
+
+    unspecified = object()
+
+    def __init__(self, input, *, args=tuple(), kwargs=None, output_process_fn_grad=None, ref=unspecified, extra=None):
+        """Parameters
+        ----------
+        input : object
+          Input is typically a single tensor that is used as the first
+          argument to op so that
+            op(input, *args, **kwargs) = input.op(*args, **kwargs) = output
+          holds. However, there exists operations that can take
+          non-Tensor objects as first arguments (e.g. many tensor
+          creation operations, torch.cat, etc) and for these cases we
+          have:
+            op(input, *args, **kwargs) = output
+        args, kwargs: tuple, dict
+          Arguments to op.
+        output_process_fn_grad: {callable, None}
+          A function that modifies the output of op compatible with input.
+        ref: {object, callable, Exception, unspecified}
+          When specified, ref determines the expected output of op for the given inputs:
+          If object, then
+            output = ref
+          If callable, then
+            output = ref(*args, **kwargs) if input is unspecified else ref(input, *args, **kwargs)
+          If Exception, then
+            try:
+                op(...)
+            except Exception as exc:
+                assert exc == ref
+        extra: {dict, None}
+          When specified, extra contains test-developer-specified information about the sample.
 
-    def __init__(self, input, *, args=tuple(), kwargs=None, output_process_fn_grad=None):
-        # input must be a single tensor, and will be the first argument to the op
-        #   this follows the typical pattern where op(t, ...) = t.op(...)
-        assert isinstance(input, torch.Tensor)
+        """
         self.input = input
         self.args = args
         self.kwargs = kwargs if kwargs is not None else {}
         self.output_process_fn_grad = output_process_fn_grad
+        self.ref = ref
+        self.extra = extra if extra is not None else {}
+        caller = inspect.getframeinfo(inspect.stack()[1][0])
+        self.extra['sample-origin'] = f'{caller.filename}#{caller.lineno}'
 
     def __repr__(self):
+        if isinstance(self.input, torch.Tensor):
```

Comment:
I'm leery or reversing coarse and allowing anything to be an input again. 

What about allowing input to be a torch.Tensor or None?

Reply:
@mruberry, consider `torch.as_tensor` operation as an example. It can take non-Tensor objects as inputs, for instance, Python sequences (lists, tuples, or any object that implements `__len__` and `__getitem__` methods), NumPy ndarrays, any objects that implement CPU or CUDA Array Interface.
Your suggestion about not allowing non-Tensors as `input` would not allow using OpInfo framework for testing tensor creation operations... unless the `None` input is interpreted as follows:
```
output = op(*args, **kwargs) if input is None else  op(input, *args, **kwargs)
```
To me, this looks like an unnecessary complication. Can you explain how the `input is None` case is useful or should be interpreted?